core/trino-main/src/main/java/io/trino/execution/DropTableTask.java

```diff
@@ -54,6 +56,28 @@ public String getName()
         Session session = stateMachine.getSession();
         QualifiedObjectName tableName = createQualifiedObjectName(session, statement, statement.getTableName());
 
+        Optional<ConnectorMaterializedViewDefinition> materializedView = metadata.getMaterializedView(session, tableName);
+        if (materializedView.isPresent()) {
+            if (!statement.isExists()) {
+                throw semanticException(
+                        TABLE_NOT_FOUND,
+                        statement,
+                        "Table '%s' does not exist, but a materialized view with that name exists. Did you mean DROP MATERIALIZED VIEW %s?", tableName, tableName);
+            }
+            return immediateVoidFuture();
+        }
+
+        Optional<ConnectorViewDefinition> view = metadata.getView(session, tableName);
+        if (view.isPresent()) {
+            if (!statement.isExists()) {
+                throw semanticException(
+                        TABLE_NOT_FOUND,
+                        statement,
+                        "Table '%s' does not exist, but a view with that name exists. Did you mean DROP VIEW %s?", tableName, tableName);
+            }
+            return immediateVoidFuture();
```

Comment:
Shouldn't we be throwing an error from here? returning a void future here means that `DROP TABLE actually_a_view` succeeds without dropping anything.

Reply:
In HiveQueryRunner:

```sql
trino:tpch> create view a_view as select * from nation;
CREATE VIEW
trino:tpch> drop table a_view;
Query 20210830_082947_00044_cki4q failed: line 1:1: Table 'hive.tpch.a_view' does not exist, but a view with that name exists. Did you mean DROP VIEW hive.tpch.a_view?
drop table a_view

trino:tpch> drop table if exists a_view;
DROP TABLE
trino:tpch> 
```

This changes the semantics - I'm not sure if this was expected or not.

Reply:
IMO this is correct behaviour since the table does not exist and `IF EXISTS` was used.